### DISCUSSION TITLE
ProgressBar Theming (incomplete)

### DISCUSSION THREAD

**Discussion by TetraTheta** [BlueMystical/Dark-Mode-Forms#29] [Ideas]
> ProgressBar: Un-themed.

ProgressBar is un-themed by Dark-Mode-Forms.

I tried to theme ProgressBar, but failed. So I decided to share what I've found so far for _future_.

![1](https://github.com/BlueMystical/Dark-Mode-Forms/assets/37061801/660bf1e2-c5f6-4745-90b6-9ebb7b297d09)

This is how ProgressBar looks like in dark mode in native. Unfortunately, MS didn't expose how to make controls dark-themed, so we have to use this screenshot as reference to replicate their looks.

# `Application.EnableVisualStyles()` blocks theming

I searched Google for how to set background color or ProgressBar.

```csharp
// ProgressBar pBar
pBar.BackColor = OScolors.Control;
SendMessage(pBar.Handle, 8193, IntPtr.Zero, 0x00FF00FF); // 8193: PBM_SETBKCOLOR
```
At least [PBM_SETBKCOLOR](https://learn.microsoft.com/en-us/windows/win32/controls/pbm-setbkcolor) should work because it is Windows API. But both didn't work.

That was because we enable visual style in `Program.cs`(main entry point) like this:
```csharp
internal static class Program {
  [STAThread]
  private static void Main() {
    Application.EnableVisualStyles(); // This line
    Application.SetCompatibleTextRenderingDefault(false);
    Application.Run(new Form1());
  }
}
```
`Application.EnableVisualStyles()` is what makes ProgressBar look like this:  
![2](https://github.com/BlueMystical/Dark-Mode-Forms/assets/37061801/5e81b593-5ee0-4421-a48c-e4712302ae08)  
Without it, we can actually theme ProgressBar as we like, but that looks very old and weird.  
![3](https://github.com/BlueMystical/Dark-Mode-Forms/assets/37061801/55ba073c-2c2b-486d-8110-d889cf3590c7)  
(Pink background color is because of `SendMessage(pBar.Handle, 8193, IntPtr.Zero, 0x00FF00FF)`)

Here is some _improved_ code, but I found that I can't change `ForeColor` of ProgressBar.
```csharp
public partial class Form1 : Form {
  [DllImport("user32.dll", CharSet = CharSet.Auto)]
  public static extern IntPtr SendMessage(IntPtr hWnd, uint msg, IntPtr wParam, IntPtr lParam);

  private IntPtr ColorToCOLORREF(Color c) {
    return (IntPtr)(c.R * 0x100 + c.G * 0x10000 + c.B);
  }

  private enum PBState : int {
    Normal = 1,
    Error = 2,
    Paused = 3,
  }

  public Form1() {
    InitializeComponent();
    var d = new DarkModeCS(this);
    var c = d.OScolors.ControlLight;
    pBar.Style = ProgressBarStyle.Continuous;

    SendMessage(pBar.Handle, 1040, (IntPtr)PBState.Error, IntPtr.Zero); // Set ProgressBar ForeColor (Doesn't work)
    SendMessage(pBar.Handle, 8193, IntPtr.Zero, ColorToCOLORREF(c)); // Set ProgressBar BackColor
  }
}
```
![4](https://github.com/BlueMystical/Dark-Mode-Forms/assets/37061801/c797220b-5ce4-4dc0-a04a-7c940dac18ae)  
Like I said, I can't change ForeColor.

I concluded that we have to override ProgressBar with new class that overrides `OnPaint(PaintEventArgs)`, with `Application.EnableVisualStyles()` enabled. I don't have time for this, so I'll leave these information here for _later use_.

**Comment by BlueMystical** [comment]
Great Info, thank you.

**Comment by BlueMystical** [comment]
Alright...
I made a FlatProgrssBar Control (on top), Bottom is default Progreesbar:
![image](https://github.com/BlueMystical/Dark-Mode-Forms/assets/10116951/17ebb9c6-b325-4609-8e69-bb1eb7d5cfca)
On Windows Clear Color mode:
![image](https://github.com/BlueMystical/Dark-Mode-Forms/assets/10116951/cd8861a6-5e81-4cc2-af27-b0f28da4d011)

it takes the backcolor of the Windows Theme.
You can choose the progress Color
Sadly the animation is missing.

I'll upload it in a few.
